rllib/offline/estimators/feature_importance.py

```diff
@@ -0,0 +1,112 @@
+        fitted predictor when the data is tablular.
+
+        This implementation is also known as permutation importance that is defined to
+        be the variation of the model's prediction when a single feature value is
+        randomly shuffled. In RLlib it is implemented as a custom OffPolicyEstimator
+        which is used to evaluate RLlib policies without performing environment
+        interactions.
+
+        Example usage: In the example below the feature importance module is used to
+        evaluate the policy and the each feature's importance is computed after each
+        training iteration. The permutation are repeated `self.repeat` times and the
+        results are averages across repeats.
+
+        ```python
+            config = (
+                AlgorithmConfig()
+                .offline_data(
+                    off_policy_estimation_methods=
+                        {
+                            "feature_importance": {
+                                "type": FeatureImportance,
+                                "repeat": 10
+                            }
+                        }
+                )
+            )
+
+            algorithm = DQN(config=config)
+            results = algorithm.train()
+        ```
+
+        Args:
+            name: string to save the feature importance results under.
+            policy: the policy to use for feature importance.
+            repeat: number of times to repeat the perturbation.
+            gamma: dummy discount factor to be passed to the super class.
+            perturb_fn: function to perturb the features. By default reshuffle the
+            features within the batch.
+        """
+        super().__init__(name, policy, gamma=gamma)
+        self.repeat = repeat
+        self.perturb_fn = perturb_fn
+
+    def estimate(self, batch: SampleBatchType) -> List[OffPolicyEstimate]:
+        """Estimate the feature importance of the policy.
+
+        Given a batch of tabular observations, the importance of each feature is
+        computed by perturbing each feature and computing the difference between the
+        perturbed policy and the reference policy. The importance is computed for each
+        feature and each perturbation is repeated `self.repeat` times.
+
+        Args:
+            batch: the batch of data to use for feature importance.
+
+        Returns:
+            A list of OffPolicyEstimate objects. Each OffPolicyEstimate object contains
+            a metics name and a dictionary of metrics mapping feature index to its
+            importance.
+        """
+
+        obs_batch = batch["obs"]
+        n_features = obs_batch.shape[-1]
+        importance = np.zeros((self.repeat, n_features))
+
+        ref_actions, _, _ = self.policy.compute_actions(obs_batch, explore=False)
+        for r in range(self.repeat):
+            for i in range(n_features):
+                copy_obs_batch = copy.deepcopy(obs_batch)
```

Comment:
@kouroshHakha Sorry I'm late to the party. Do you think it would be worth the investigation to see how much time we would save if we would not copy the whole batch but rather the column to be permuted?

Reply:
Maybe test this with a large feature vector and see what happens?

Reply:
Obviously this would alter the steps here a little. "copy -> permute original -> compute error -> reinsert copy"

Reply:
Your point is valid. This can definitely be optimized further. We can worry about it when it becomes a bottleneck. I would say it is a P2 for now.